Add <Route strict>

rafrex opened the pull request:
A fix for #4170 

Can now add a `strict` boolean prop to `Route` (if the strict prop is omitted it defaults to `true`):
```js
// this will match both /foo and /foo/
<Route exact path="/foo" strict={false} component={foo} />
```

I ran the tests and it passes all of them, and in my limited testing it works both inside and outside of `<Switch />`

mjackson commented:
Thanks, @rafrex :) Right now our automated tests on v4 are not very complete, unfortunately. Would you mind booting the website and clicking through the examples to make sure everything still works?

rafrex commented:
I clicked through all the examples and they seem to all work.

Note that I locally changed the default for `strict` to false in the router, and then booted the site so I  wouldn't have to add `strict={false}` to every route in the examples. I also tested a few routes by adding the `strict` prop as well.

simonkberg commented:
Curious, how come `strict` is enabled by default? 
Not matching trailing slashes seems like the edge case here, which would be in line with the default behavior of `path-to-regex` (which mimics express' behavior).

mjackson commented:
Sounds good, @rafrex. Can we get a few tests around this too? Ideally we'd have 2 tests: one for `strict = false` and another for `strict = true`.

rafrex commented:
@simenbrekken I left the default as per how it worked previously. I agree it might make sense to change it to `false`.

Phoenixmatrix commented:
Just a nitpicky 2 cents: the fact that the internal implementation (the path-to-regexp) uses an argument called "strict" is kind of an implementation detail, and that terminology isn't super intuitive for a user of the router. (like, wtf is "strict...oh, its a param for an internal dependency!".

if that gets exposed as a public api, it probably needs a better name.

rafrex commented:
I actually don't mind the `strict` prop name because RRv4 doc says refer to `path-to-regex` to figure out how `path="..."` will match. But also don't have a strong preference for it either.

Maybe the default could be `false` and the prop could be `matchTrailingSlash` or something.

@mjackson what do think the default for `strict` should be? And what do you think the prop should be called? (I think both of these are high level api decisions).

mjackson commented:
I think `strict` is a fine name and we should probably use the same default as `path-to-regexp` does. I'm sure those folks have good reasons by now for choosing the defaults they have.

rafrex commented:
Change the default for strict to `false`.

I don't have experience with karma/mocha, someone with more experience want to write the tests?

mjackson commented:
All you have to do is copy/paste some of the existing tests. Look at `integration-test.js` for hints.

mjackson commented:
Also, we need to make sure we have tests for cases like https://github.com/ReactTraining/react-router/issues/4130 where two routes have similar URLs. @SimeonC created https://github.com/ReactTraining/react-router/pull/4133 with some tests that we may be able to use here (with attribution, ofc).

SimeonC commented:
I'll just reply here - I can't update/create PRs currently as I'm without a PC for a month or two while moving so feel free to repurpose any of my PRs as needed.

rafrex commented:
@mjackson I added tests for the strict prop with trailing slash.

I also added tests for similar paths (`/some` and `/some-path`). The problem in #4130 was with active links, which are no longer part of v4 (as far as I can tell), so I tested it using `children` and making sure the `match` boolean is correct (which seems to be how one would do active links now).

```js
<Route path="/some-path" children={({ match }) => (
  <h1>{match ? TEXT2 : TEXT1}</h1>
)}/>
```

mjackson commented:
Thanks, @rafrex! Looks like there are some conflicts here, probably because I moved some stuff around when organizing the monorepo last week. Would you please rebase?

Also, sorry for all the churn. I'm trying to get things to settle down as quickly as I can.

rafrex commented:
@mjackson I rebased, should be all set!

pshrmn commented on packages/react-router/modules/__tests__/integration-test.js:
```diff
@@ -38,6 +38,79 @@ describe('Integration Tests', () => {
     expect(div.innerHTML).toNotContain(TEXT)
   })
 
+  it('renders an exact route with trailing slash', () => {
```
Should there also be a test of the inverse? `<Route path='/somepath/'>` when `location.pathname = '/somepath'`?

rafrex replied:
Sure, why not - added.

timdorr commented on packages/react-router/modules/Core.js:
```diff
@@ -1,18 +1,18 @@
 import pathToRegexp from 'path-to-regexp'
 
-const patternCache = { true: {}, false: {} }
+const patternCache = { truetrue: {}, truefalse: {}, falsefalse: {}, falsetrue: {} }
```
This is fine for now, but put a pin here to refactor this relative nonsensical code soon. We're bound to sprout a bug here eventually 🙃

mjackson commented:
Thanks, @rafrex. I'll take it from here ;)
